src/Elastic.Apm.Azure.Functions/ElasticApmMiddleware.cs

```diff
@@ -0,0 +1,126 @@
+// Licensed to Elasticsearch B.V under
+// one or more agreements.
+// Elasticsearch B.V licenses this file to you under the Apache 2.0 License.
+// See the LICENSE file in the project root for more information
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Elastic.Apm.Api;
+using Elastic.Apm.Cloud;
+using Elastic.Apm.Helpers;
+using Elastic.Apm.Logging;
+using Elastic.Apm.Model;
+using Microsoft.Azure.Functions.Worker;
+using Microsoft.Azure.Functions.Worker.Http;
+using Microsoft.Azure.Functions.Worker.Middleware;
+using TraceContext = Elastic.Apm.DistributedTracing.TraceContext;
+
+namespace Elastic.Apm.Azure.Functions;
+
+public class ElasticApmMiddleware : IFunctionsWorkerMiddleware
+{
+	private static readonly IApmLogger Logger;
+	private static readonly string FaasIdPrefix;
+	private static int ColdStart = 1;
+
+	static ElasticApmMiddleware()
+	{
+		Logger = Agent.Instance.Logger.Scoped(nameof(ElasticApmMiddleware));
+		var metaData = new AzureFunctionsMetadataProvider(Logger,
+			new EnvironmentVariables(Logger).GetEnvironmentVariables()).GetMetadata();
+		FaasIdPrefix =
+			$"/subscriptions/{metaData?.Account?.Id}/resourceGroups/{metaData?.Project}/providers/Microsoft.Web/sites/{metaData?.Instance}/functions/";
+		Logger.Trace()?.Log($"FaasIdPrefix: {FaasIdPrefix}");
+	}
+
+	private static bool IsColdStart() => Interlocked.Exchange(ref ColdStart, 0) == 1;
+
+	public async Task Invoke(FunctionContext context, FunctionExecutionDelegate next)
+	{
+		try
+		{
+			Logger.Trace()?.Log($"{nameof(Invoke)} - {context.FunctionDefinition.Name}");
+
+			var data = GetTriggerSpecificData(context);
+			await Agent.Tracer.CaptureTransaction(data.Name, ApiConstants.TypeRequest, async t =>
+			{
+				if (t is Transaction transaction) // TODO: Expose FaaS property on ITransaction?
```

Comment:
> // TODO: Expose FaaS property on ITransaction?

Yeah, I'd do it in this PR. The `Faas` type is already public, so let's properly expose it - it's better for us (no need to cast here) and better for users (we don't expose public types that they can't access on our interfaces and use e.g. in filters).